sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala

```diff
@@ -317,7 +363,74 @@ object RewriteDistinctAggregates extends Rule[LogicalPlan] {
       }
       Aggregate(groupByAttrs, patchedAggExpressions, firstAggregate)
     } else {
-      a
+      val (distinctAggExpressions, regularAggExpressions) = aggExpressions.partition(_.isDistinct)
+      if (distinctAggExpressions.exists(_.filter.isDefined)) {
+        val regularAggExprs = regularAggExpressions.filter(e => e.children.exists(!_.foldable))
+        val regularFunChildren = regularAggExprs
+          .flatMap(_.aggregateFunction.children.filter(!_.foldable))
+        val regularFilterAttrs = regularAggExprs.flatMap(_.filterAttributes)
+        val regularAggChildren = (regularFunChildren ++ regularFilterAttrs).distinct
+        val regularAggChildAttrMap = regularAggChildren.map(expressionAttributePair)
+        val regularAggChildAttrLookup = regularAggChildAttrMap.toMap
+        val regularOperatorMap = regularAggExprs.map {
+          case ae @ AggregateExpression(af, _, _, filter, _) =>
+            val newChildren = af.children.map(c => regularAggChildAttrLookup.getOrElse(c, c))
+            val raf = af.withNewChildren(newChildren).asInstanceOf[AggregateFunction]
+            val filterOpt = filter.map(_.transform {
+              case a: Attribute => regularAggChildAttrLookup.getOrElse(a, a)
+            })
+            val aggExpr = ae.copy(aggregateFunction = raf, filter = filterOpt)
+            (ae, aggExpr)
+        }
+        val distinctAggExprs = distinctAggExpressions.filter(e => e.children.exists(!_.foldable))
+        val rewriteDistinctOperatorMap = distinctAggExprs.zipWithIndex.map {
+          case (ae @ AggregateExpression(af, _, _, filter, _), i) =>
+            val phantomId = i + 1
```

Comment:
I think the current logic is complicated for reviews... e.g., what's the purpose of `phantomId`?

Reply:
In the current approach, the more distinct aggregates they are, the more rows `Expand` generates? It's like `select count(distinct a), count(distinct a) filter a > 3, count(distinct a) filter a < 1, ...`

Reply:
// Why do we need to construct the phantom id ?
            // First, In order to reduce costs, it is better to handle the filter clause locally.
            // e.g. COUNT (DISTINCT a) FILTER (WHERE id > 1), evaluate expression
            // If(id > 1) 'a else null first, and use the result as output.
            // Second, If more than one DISTINCT aggregate expression uses the same column,
            // We need to construct the phantom attributes so as the output not lost.
            // e.g. SUM (DISTINCT a), COUNT (DISTINCT a) FILTER (WHERE id > 1) will output
            // attribute 'a and attribute 'phantom1-a instead of two 'a.
            // Note: We just need to illusion the expression with filter clause.

Reply:
Why we need to process multiple groups (grouped by phantom keys) for the same distinct group (e.g., `ount(distinct a) filter where a > 3` and `count(distinct a) filter where a < 2`)?

Reply:
The illusionary mechanism may result in multiple distinct aggregations uses different column, so we still need to call `rewrite`. (e.g., count(distinct phantom1-a) and count(distinct a phantom2-a))

Reply:
In the current approach, it seems filter conditions are evaluated in not aggregate but expand. Even in this case (the same aggregate group), we cannot use the aggregate filter logic in `AggregationIterator` implemented in your previous commit?

Reply:
You can reference the discussion
https://github.com/apache/spark/pull/26656#discussion_r353717031
In order to reduce cost, we treat the filter conditions of DISTINCT aggregate expr in first aggregate is better.
If we use `AggregationIterator`, the filter conditions of DISTINCT aggregate expr will be treated in second or thrid aggregate.